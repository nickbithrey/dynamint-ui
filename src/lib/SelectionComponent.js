import React from 'react';
import PropTypes from 'prop-types';
import Table from '~/lib/Table';
import { Link } from 'react-router-dom';
import Button from '~/lib/Button';
import { Stack } from 'office-ui-fabric-react/lib/Stack';
import { MarqueeSelection } from 'office-ui-fabric-react/lib/MarqueeSelection';
import { Selection } from 'office-ui-fabric-react/lib/DetailsList';

export default class SelectionComponent extends React.Component {
	
	constructor(props) {
		super(props);
		this.state = {};

		this.selection = new Selection({
		    onSelectionChanged: () => {
		        this.setState({
		            selectionDetails: this.getSelectionDetails()
		        });
		    }
	    });
		this.renderButton = this.renderButton.bind(this);
	}
	
	getSelectionDetails() {
		if (this.selection.getSelectedCount() > 0) {
			return this.selection;
		}
	}
	
	renderTitle(title) {
		if (title) {
			return (
				<div>
					<h1>{title}</h1>
				</div>
			);
		}
	}

	renderSelectionTable(list, loadList, columns) {
		return (
			<MarqueeSelection selection={this.selection}>
				<Table list={list} load={loadList} columns={columns} selection={this.selection}></Table>
			</MarqueeSelection>
		);
	}

	renderButtons(buttons = []) {
		return buttons.map(btn => btn.build()).filter(btn => btn.condition(this.selection)).map(this.renderButton);
	}
	
	renderButton(btn, i) {
		let to = {
			pathname: btn.pathname,
			state: btn.stateFn(this.selection)
		};
		return (
			<Link key={i} to={to}><Button text={btn.text} /></Link>
		);
	}
	
	render() {
		const {
			title, 
			list, 
			loadList, 
			columns, 
			buttons
		} = this.props;
		return (
			<div>
				{this.renderTitle(title)}
				{this.renderSelectionTable(list, loadList, columns)}
				<Stack horizontalAlign={'end'}>
					<Stack horizontal tokens={{childrenGap: 5}} horizontal-align="space-between">
						{this.renderButtons(buttons)}
					</Stack>
				</Stack>
			</div>
		);
	}
	
}

SelectionComponent.propTypes = {
	title: PropTypes.string,
	list: PropTypes.array,
	loadList: PropTypes.func,
	columns: PropTypes.arrayOf(PropTypes.object).isRequired,
	buttons: PropTypes.arrayOf(PropTypes.object)
};

export class SelectionButton {
	
	constructor(text, path, condition, stateFn) {
		this.text = text;
		this.path = path;
		this.condition = condition;
		this.stateFn = stateFn;
	}
	
	build() {
		return {
			text: this.text,
			pathname: this.path,
			stateFn: this.stateFn ? this.stateFn : () => {},
			condition: this.condition ? this.condition : alwaysShow
		};
	}
	
}

const alwaysShow = () => true;

const hasSelection = selection => selection.getSelectedCount() > 0;

export {
	alwaysShow,
	hasSelection
};
