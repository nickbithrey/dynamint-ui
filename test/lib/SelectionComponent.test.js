import React from 'react';
import { shallow } from 'enzyme';

import SelectionComponent, { SelectionButton } from '~/lib/SelectionComponent';

describe('<SelectionComponent /> rendering', () => {

	it('can render standard component with records', () => {
		const props = {
			title: 'titleText',
			list: [],
			loadList: jest.fn(() => []),
			columns: [],
			buttons: [
					new SelectionButton('Edit', 'path', jest.fn(sel => true), jest.fn(sel => {return {}})), 
					new SelectionButton('Create', 'path', jest.fn(sel => true), jest.fn(sel => {return {}}))
				]
		};
		const wrapper = shallow( <SelectionComponent {...props} /> );
		
		expect(wrapper).toMatchSnapshot();
	});
		
	it('only renders buttons that have true conditions', () => {
		const props = {
			title: 'titleText',
			list: [],
			loadList: jest.fn(() => []),
			columns: [],
			buttons: [
					new SelectionButton('Edit', 'path', jest.fn(sel => false), jest.fn(() => {return {}})), 
					new SelectionButton('Create', 'path', jest.fn(sel => true), jest.fn(() => {return {}}))
				]
		};
		const wrapper = shallow( <SelectionComponent {...props} /> );

		const btn = wrapper.find('Button');
		expect(btn.length).toBe(1);
		expect(btn.prop('text')).toBe('Create');
		expect(props.buttons[0].condition.mock.calls).toHaveLength(1);
		expect(props.buttons[0].stateFn.mock.calls).toHaveLength(0);
		expect(props.buttons[1].condition.mock.calls).toHaveLength(1);
		expect(props.buttons[1].stateFn.mock.calls).toHaveLength(1);
	});
	
	it('renders nothing for when no title is supplied', () => {
		const props = {
			title: 'titleText',
			list: [],
			loadList: jest.fn(() => []),
			columns: [],
			buttons: [
					new SelectionButton('Edit', 'path', jest.fn(sel => true), jest.fn(() => {return {}})), 
					new SelectionButton('Create', 'path', jest.fn(sel => true), jest.fn(() => {return {}}))
				]
		};
		const wrapper = shallow( <SelectionComponent {...props} /> );

		let title = wrapper.find('h1');
		expect(title.length).toBe(1);
		expect(title.text()).toBe(props.title);
		
		wrapper.setProps({title: null});
		
		title = wrapper.find('h1');
		expect(title.length).toBe(0);
	});
		
	it('renders table with details', () => {
		const props = {
			title: 'titleText',
			list: [
					{text: 'val1', num: 1},
					{text: 'val2', num: 2}
				],
			loadList: jest.fn(() => []),
			columns: [
					{fieldName: 'text', name: 'Text'},
					{fieldName: 'num', name: 'Number'}
				],
			buttons: [
					new SelectionButton('Edit', 'path', jest.fn(sel => true), jest.fn(() => {return {}})), 
					new SelectionButton('Create', 'path', jest.fn(sel => true), jest.fn(() => {return {}}))
				]
		};
		const wrapper = shallow( <SelectionComponent {...props} /> );

		let table = wrapper.find('Table');
		expect(table.length).toBe(1);
		expect(table.prop('list')).toBe(props.list);
		expect(table.prop('load')).toBe(props.loadList);
		expect(table.prop('columns')).toBe(props.columns);
	});
	
	it('re-runs button condition when selecting record in table', () => {
		const props = {
			title: 'titleText',
			list: [
					{text: 'val1', num: 1},
					{text: 'val2', num: 2}
				],
			loadList: jest.fn(() => []),
			columns: [
					{fieldName: 'text', name: 'Text'},
					{fieldName: 'num', name: 'Number'}
				],
			buttons: [
					new SelectionButton('Edit', 'path', jest.fn(sel => sel.getSelectedCount() > 0), jest.fn(() => {return {}})), 
					new SelectionButton('Create', 'path', jest.fn(sel => true), jest.fn(() => {return {}}))
				]
		};
		const wrapper = shallow( <SelectionComponent {...props} /> );

		let btn = wrapper.find('Button');
		expect(btn.length).toBe(1);
		expect(btn.prop('text')).toBe('Create');

		// call selection changed event
		const selection = wrapper.find('StyledMarqueeSelectionBase').props().selection;
		selection._items = props.list;
		selection.setIndexSelected(1, true, true);
		selection._onSelectionChanged();
		
		btn = wrapper.find('Button');
		expect(btn.length).toBe(2);
		expect(btn.at(0).prop('text')).toBe('Edit');
		expect(btn.at(1).prop('text')).toBe('Create');
	});
	
});