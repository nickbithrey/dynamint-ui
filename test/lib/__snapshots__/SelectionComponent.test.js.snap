// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectionComponent /> rendering can render standard component with records 1`] = `
<div>
  <div>
    <h1>
      titleText
    </h1>
  </div>
  <Rowise
    tokens={
      Object {
        "childrenGap": 10,
      }
    }
  >
    <Block>
      <StyledMarqueeSelectionBase
        selection={
          Selection {
            "_anchoredIndex": 0,
            "_canSelectItem": [Function],
            "_changeEventSuppressionCount": 0,
            "_exemptedCount": 0,
            "_exemptedIndices": Object {},
            "_getKey": [Function],
            "_isModal": false,
            "_items": Array [],
            "_keyToIndexMap": Object {},
            "_onSelectionChanged": [Function],
            "_selectedItems": null,
            "_unselectableCount": 0,
            "_unselectableIndices": Object {},
            "count": 0,
            "mode": 2,
          }
        }
      >
        <Table
          columns={Array []}
          list={Array []}
          load={[MockFunction]}
          selection={
            Selection {
              "_anchoredIndex": 0,
              "_canSelectItem": [Function],
              "_changeEventSuppressionCount": 0,
              "_exemptedCount": 0,
              "_exemptedIndices": Object {},
              "_getKey": [Function],
              "_isModal": false,
              "_items": Array [],
              "_keyToIndexMap": Object {},
              "_onSelectionChanged": [Function],
              "_selectedItems": null,
              "_unselectableCount": 0,
              "_unselectableIndices": Object {},
              "count": 0,
              "mode": 2,
            }
          }
        />
      </StyledMarqueeSelectionBase>
    </Block>
    <Block>
      <Columnar
        horizontalAlign="end"
        tokens={
          Object {
            "childrenGap": 5,
          }
        }
      >
        <Block
          key="0"
        >
          <Link
            replace={false}
            to={
              Object {
                "pathname": "path",
                "state": Object {},
              }
            }
          >
            <Button
              text="Edit"
            />
          </Link>
        </Block>
        <Block
          key="1"
        >
          <Link
            replace={false}
            to={
              Object {
                "pathname": "path",
                "state": Object {},
              }
            }
          >
            <Button
              text="Create"
            />
          </Link>
        </Block>
      </Columnar>
    </Block>
  </Rowise>
</div>
`;
